src/compiler/crystal/syntax/lexer.cr

```diff
@@ -127,64 +127,35 @@ module Crystal
             next_char :"<<="
           when '-'
             here = StringIO.new(20)
-            here_start = 0
 
             while true
               case char = next_char
               when '\n'
                 @line_number += 1
                 @column_number = 0
-                next_char
-                here_start = current_pos
                 break
+              when '\\'
+                if peek_next_char == 'n'
+                  next_char
+                  raise "invalid heredoc identifier"
+                end
+              when ' '
+                case peek_next_char
+                when ' '
+                  next_char
+                when '\n'
+                  next_char
+                  break
+                else
+                  raise "invalid heredoc identifier"
+                end
```

Comment:
1. Why is the `\n` case needed? 
2. Why are the double space or space newline cases needed?